Implement `cluster-setup` command to deploy needed charts

detjensrobert opened the pull request:
Adds a new command, `cluster-setup` to install cluster software dependencies. 

Resolves #23.

Currently, this installs the following charts to the `ingress` namespace:
- [ingress-nginx](https://kubernetes.github.io/ingress-nginx/)
- [cert-manager](https://cert-manager.io/docs/installation/helm/) (with LetsEncrypt prod/staging issuers)
- [external-dns](https://github.com/bitnami/charts/tree/main/bitnami/external-dns) (unconfigured)

These are *not* deployed with Helm clientside, as there is no Helm API binding for Rust (only Go) and I assume we do not want to shell out to the Helm CLI. Instead, this uses an Rancher RKE/K3D project `helm-controller` ([source](https://github.com/k3s-io/helm-controller) and [docs](https://docs.k3s.io/helm)) that watches for charts as CRDs and deploys them. This seems to work pretty well, and lets us split up the ingress chart values between 'global' settings and the tcp port expose configs.

The only pain point with this is getting the status of the deployment, since we're not running the deploy clientside we can't read the output/return code of helm. We need to get the job name from the chart CRD, and watch the status of that. I don't have that implemented yet. 

External DNS chart may also be out of scope, since this needs a lot of different configuration depending on what dns provider clients are using. I think we might want to leave this to the user or otherwise discuss how to handle how to pass credentials in.

detjensrobert committed:
Move kube and docker clients to shared module

since these are now used in multiple places make em reusable

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
move docker client error conversion to shared method

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
add partial cluster-setup, only helm operator for now

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Install ingress, cert-manager, and external-dns in cluster setup

these are all installed to the `ingress` namespace. Cert-manager's email
is not set and external-dns has no creds. This is TBD on how to expose
these to the user (if at all -- external-dns may be out of scope)

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
TODO for cert-manager issuer emails

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
move all template and manifest files to new asset folder

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
deploy helm charts for ingress certmanager and dns

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
fix chart/repo url for external-dns chart

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
split up kube resource and client helper methods

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
set custom ingress class for our managed ingress-nginx

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
clippy fixed

Signed-off-by: Robert Detjens <github@detjens.dev>

cacama-valvata reviewed (changes requested):
results of conversation on 01/12: 
- shell out for the helm api since there is no threat-model even in the presence of user (infra admin) input
  - use the safe rust "shelling out" utilities that you mentioned during meeting
- for external-dns:
  -  the easy, acceptable-for-short-term solution can be the passed-along string that you mentioned
  - long term should be separate structs in the configs for the different providers

detjensrobert committed:
Add opaque dns field in profile config for external dns config

We don't need to parse whatever the user puts in there, since it will
be passed verbatim into the chart values for external-dns.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Switch to minijinja for templating

This crate is much smaller and simpler and more in line with what we're using
it for (one-off static templating). Also, it pulls in less dependencies.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Allow configuring external-dns chart via profile config field

Since the external-dns chart configuration will differ greatly between
providers, we cannot know what options the user needs to set. Instead,
have them provide whatever config is needed and template that into the
external-dns chart values.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Ignore extracted build artifact files

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Switch bundled setup charts from full CRD to only values

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Switch to k3d since minikube has problems with cert-manager

for some reason, minikube running rootless cannot extract the images used
by cert-manager (tar permission denied). k3d does not have this problem
and is faster to set up anyway.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Switch to shelling out to helm instead of helm-operator

This is a lot easier to report errors on and log any output from helm.
Since we are not taking user input for commands, this is safe and does
not pose a security risk like shelling out normally could have.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
fix ingress class name

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
expose ingress ports in k3d

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
clippy loop fix

Signed-off-by: Robert Detjens <github@detjens.dev>

KekoaM reviewed (approved):
Overall changes look good. 
Potentially will need more testing for external dns once we are able to deploy to an actual cluster
